python/hidet/transforms/lower_integer_subbyte.py

```diff
@@ -0,0 +1,349 @@
+
+
+class LowerIntegerSubbyteRewriter(IRRewriter):
+    # convert subbyte integers to their storage type
+    # e.g.,
+    # a = register_tensor("int4b", [16]) ==> a = register_tensor("uint8", [8])
+    # int4b* ptr = &a[8]                 ==> uint8* ptr = &a[4]
+    # int4b* ptr = ptr + 10              ==> uint8* ptr = ptr + 5
+    # we support tensor element access for global|shared|register tensors, and
+    # we support buffer store statment for register tensors, i.e.
+    # a = register_tensor("int4b", [4, 4])
+    # b = a[2, 2]
+    # a[2, 2] = int4b(-5)
+    # user has to explicitly insert dtype conversion before applying arithmetic
+    # operations on subbyte integers.
+    # int4b a = int4b(-2)
+    # int4b b = int4b(-3)
+    # int4b c = a + b                    ==> not allowed
+    def __init__(self):
+        super().__init__()
+        self.old2new: Dict[Var, Var] = {}
+        self.stmts: List[Stmt] = []
+        self.var2scope: Dict[Var, DeclareScope] = {}
+        self.recursive_depth = 0
+
+    def auto_var(self, v: Var = None, hint: str = None, e: Expr = None):
+        if v is not None:
+            self.stmts.append(DeclareStmt(v))
+            return v
+        v_ty = infer_type(e)
+        v = var(hint, v_ty)
+        self.stmts.append(DeclareStmt(v, e))
+        return v
+
+    def append_stmt(self, stmt: Union[Stmt, Expr]):
+        if isinstance(stmt, Expr):
+            stmt = EvaluateStmt(stmt)
+        self.stmts.append(stmt)
+
+    def flush_stmts(self):
+        stmts = self.stmts
+        self.stmts = []
+        return stmts
+
+    def flatten_stmts(self, stmts: List[Stmt]):
+        if len(stmts) == 1:
+            return stmts[0]
+        else:
+            return SeqStmt(stmts)
+
+    def _get_divisor(self, dtype: DataType):
+        storage_ty = dtype.storage
+        storage_bits = storage_ty.nbits
+        dtype_bits = dtype.nbits
+        divisor = storage_bits // dtype_bits
+        return divisor
+
+    def _get_subbyte_value(self, dtype: DataType, base: Var, offset: Expr):
+        storage_ty = dtype.storage
+        storage_bits = storage_ty.nbits
+        dtype_bits = dtype.nbits
+        divisor = storage_bits // dtype_bits
+        if divisor & (divisor - 1) != 0:
+            raise TypeError(f"data type not supported yet(got:{dtype})")
+        idx = simplify(offset // divisor)
+        offset_ = simplify(offset & (divisor - 1))
+        mask = storage_ty.constant(dtype.bits_mask)
+        return (base[idx] >> (offset_ * dtype_bits)) & mask
+
+    def _set_subbyte_value(self, dtype: DataType, base: Var, offset: Expr, value: Expr):
+        storage_ty = dtype.storage
+        storage_bits = storage_ty.nbits
+        dtype_bits = dtype.nbits
+        divisor = storage_bits // dtype_bits
+        if divisor & (divisor - 1) != 0:
+            raise TypeError(f"data type not supported yet(got:{dtype})")
+        idx = simplify(offset // divisor)
+        offset_ = simplify(offset & (divisor - 1))
```

Comment:
I usually write `offset % divisor` directly and let the compiler to perform the optimization. But it's okay if you find this is more readable or comfortable to you.

Reply:
done